geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/Z3IdxStrategy.scala

```diff
@@ -115,7 +115,15 @@ class Z3IdxStrategy extends Strategy with Logging with IndexFilterHelpers  {
     val epochWeekStart = Weeks.weeks(week)
     val prefix = Shorts.toByteArray(epochWeekStart.getWeeks.toShort)
 
-    val z3ranges = Z3_CURVE.ranges(lx, ly, ux, uy, lt, ut, 8)
+    // lazily try to find the min recursion that will give us ranges
+    // if recursion is too high, big ranges will be expensive to compute
+    // if too low then small ranges will not return anything
+    val z3rangeCandidates = Z3_RECURSION_TIERS.iterator.map(Z3_CURVE.ranges(lx, ly, ux, uy, lt, ut, _))
```

Comment:
I think @jnh5y has a stop-gap fix that doesn't require recomputing the ranges.  After that, we need to do something a bit more comprehensive to determine recursion depth based on query region.